docs/source/contribute.rst

```diff
@@ -0,0 +1,327 @@
+This guide explains how to contribute to cocotb, and documents the processes we agreed on to manage the project.
+All processes in this document are designed to streamline the development effort, to avoid bottlenecks, and to ultimately give a pleasant experience to all involved.
+
+Feeling lost?
+=============
+
+Cocotb is a diverse and challenging project to contribute to.
+If you ever feel lost, out of your depth, or simply want to know more, the `cocotb Gitter channel <https://gitter.im/cocotb/Lobby>`__ is actively watched by many cocotb users, contributors, and maintainers.
+It is a good idea if you are unsure whether your problem or question is worthy of a Github Issue to first post it to the Gitter channel.
+You may also ask questions in `Github issues <https://github.com/cocotb/cocotb/issues>`_.
+If you don't receive any response on the Gitter channel or a Github issue, or you want help privately, you may directly contact a maintainer.
+
+Architecture and Scope of Cocotb
+================================
+
+Cocotb has seen adoption in a wide variety of scenarios with sometimes conflicting requirements.
+To foster experimentation and to decentralize the development process the architecture of cocotb is highly modular.
+A solid core forms the foundation upon which extensions can provide higher-level functionality.
+
+The core of cocotb are
+
+- the infrastructure to write testbenches with coroutines, threads, etc.,
+- the abstraction and interaction with simulators through interfaces like VPI, GPI, etc.,
+- tooling to run tests, and
+- core primitives to interact with the simulation: triggers, data access classes, etc.
+
+As a general rule, functionality beyond this core set should go into extensions.
+However, none of these rules are set in stone.
+They can and should be challenged at times to ensure the project stays relevant to the majority of its users.
+
+What to Work On
+===============
+
+There is *a lot* of work to do on this project, no matter your area of expertise or skill level.
+
+If you are a beginner there are several `Github issues <https://github.com/cocotb/cocotb/issues>`__ marked `good first issue <https://github.com/cocotb/cocotb/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22>`__ you can look at.
+There are also a number of things you can work on that aren't in Github issues and don't require in-depth knowledge of the cocotb internals.
+They including the following:
+
+-  Documentation improvements
+-  Managing Github issues and the Gitter channel
+-  Testing and coverage improvements
+
+Cocotb is still not perfect.
+There are plenty of `bug fixes <https://github.com/cocotb/cocotb/issues?q=is%3Aopen+is%3Aissue+label%3Atype%3Abug>`__ and `features <https://github.com/cocotb/cocotb/issues?q=is%3Aopen+is%3Aissue+label%3Atype%3Afeature>`__ that can be worked on.
+Most of these are recorded as Github issues.
+
+Documentation
+-------------
```

Comment:
These items were originally on the same heading level as "What to work on"; I pushed them down one hierarchy level to give better structure to the document.